:bug: [JENKINS-75171] Fix ambiguous 12AM/12PM time format displayed on Build History Widget

SheikhSuhail19 opened the pull request:
This PR resolves [JENKINS-75171](https://issues.jenkins.io/browse/JENKINS-75171) by adding new `getDisambiguatedTime` method to correctly display "12:00 midnight" for 12AM and "12:00 noon" for 12PM to avoid confusion.

### Before
- 12:00PM → 12:00 PM (ambiguous)
- 12:00AM → 12:00 AM (ambiguous)

### After
- 12:00PM → 12:00 noon ✅
- 12:00AM → 12:00 midnight ✅

<!-- Comment:
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue ID you created in Jira.
Note that if you want your changes backported into LTS, you need to create a Jira issue. See https://www.jenkins.io/download/lts/#backporting-process for more information.
-->

See [JENKINS-75171](https://issues.jenkins.io/browse/JENKINS-75171).

<!-- Comment:
If the issue is not fully described in Jira, add more information here (justification, pull request links, etc.).

 * We do not require Jira issues for minor improvements.
 * Bug fixes should have a Jira issue to facilitate the backporting process.
 * Major new features should have a Jira issue.
-->

### Testing done

- Added unit tests in `FunctionsTest.java` to cover all edge cases related to disambiguating time strings, including:
  - Normal AM/PM times (e.g., `9:15 AM`)
  - Noon (`12:00 noon`)
  - Midnight (`12:00 midnight`)
  - 24-hour formatted times (e.g., `13:30`, `23:59`)
  - One second after midnight (`12:00:01 AM`)
  - One second after noon (`12:00:01 PM`)
  - Time in different time zones (e.g., converting `12:00 UTC` to a different local time zone)
  - Handling of Daylight Saving Time (e.g., transitioning from 2:00 AM to 3:00 AM)
- Verified that all new and existing unit tests pass locally using `mvn test`
- Deployed the changes on a local Jenkins instance and verified that time values are correctly disambiguated in the UI
- Confirmed correct rendering and labeling of edge cases like `12:00 midnight` and `12:00 noon`

## Before
![12am](https://github.com/user-attachments/assets/93bc19e4-f2d6-4335-90e2-8c86dd0d30a2)

## After
![1200midnight](https://github.com/user-attachments/assets/a7c6ef44-627e-44d3-b2c2-69bb575bcd99)
![1200noon](https://github.com/user-attachments/assets/4cc4751e-0bad-4615-85cb-685adf9799c2)

![image](https://github.com/user-attachments/assets/75ef7355-fc15-410d-ac6c-7ec05e09e9e4)


### Proposed changelog entries

- Bug Fix: Disambiguate 12:00 AM and 12:00 PM in human-readable time format to display as “12:00 midnight” and “12:00 noon on the Build History Widget”.
- 
<!-- Comment:
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/

Do not include the Jira issue in the changelog entry.
Include the Jira issue in the description of the pull request so that the changelog generator can find it and include it in the generated changelog.

You may add multiple changelog entries if applicable by adding a new entry to the list, e.g.
- First changelog entry
- Second changelog entry
-->

### Proposed changelog category

/label bug, web-ui

<!--
The changelog entry needs to have a category which is selected based on the label.
If there's no changelog then the label should be `skip-changelog`.

The available categories are:
* bug - Minor bug. Will be listed after features
* developer - Changes which impact plugin developers
* dependencies - Pull requests that update a dependency
* internal - Internal only change, not user facing
* into-lts - Changes that are backported to the LTS baseline
* localization - Updates localization files
* major-bug - Major bug. Will be highlighted on the top of the changelog
* major-rfe - Major enhancement. Will be highlighted on the top
* rfe - Minor enhancement
* regression-fix - Fixes a regression in one of the previous Jenkins releases
* removed - Removes a feature or a public API
* skip-changelog - Should not be shown in the changelog

Non-changelog categories:
* web-ui - Changes in the web UI

Non-changelog categories require a changelog category but should be used if applicable,
comma separate to provide multiple categories in the label command.
-->

### Proposed upgrade guidelines

N/A

<!-- Comment:
Leave the proposed upgrade guidelines in the pull request with the "N/A" value if no upgrade guidelines are needed.
The changelog generator relies on the presence of the upgrade guidelines section as part of its data extraction process.
-->

### Submitter checklist

- [x] The Jira issue, JENKINS-75171, is well-described and linked in the PR description.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change and are in the imperative mood.
- [x] There is automated testing and visual verification by deploying and validating the UI.
- [x] No new public classes, fields, or methods introduced that require `@Restricted` or `@since TODO` annotations.
- [x] No new deprecations introduced.
- [x] No new JavaScript changes or use of `eval`.
- [x] No dependency updates in this change.
- [x] No new APIs or extension points introduced.

### Desired reviewers

@jtnord
@jglick 
@alanharder 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

Commit:
:bug: [JENKINS-75171] Fix ambiguous 12AM/12PM time format displayed on Build History Widget

Add new `getDisambiguatedTime` method to correctly display "12:00 midnight" for 12AM and "12:00 noon" for 12PM to avoid confusion. Also add unit tests to verify formatting for standard times, noon and midnight cases.

jtnord commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -736,6 +736,44 @@ public static String getUserTimeZone() {
+        int hour = calendar.get(Calendar.HOUR_OF_DAY);
+        int minute = calendar.get(Calendar.MINUTE);
+
+        if (minute == 0) {
+            if (hour == 0) {
+                return Messages.DisambiguatedTime_Midnight();
+            } else if (hour == 12) {
+                return Messages.DisambiguatedTime_Noon();
+            }
+        }
```
you only wan tto do this with TimeZones that do not use the 24 hour clock.  anything using the 24 hour clock (most of Europe for example) should say 12:00 as that is always noon (00:00 would be midnight and there is no confusion between the two).

SheikhSuhail19 replied:
Thank you for your insightful feedback. I have updated the implementation to conditionally replace 'AM'/'PM' with 'Midnight'/'Noon' only for locales utilizing the 12-hour clock format. For locales that employ the 24-hour clock, the time will continue to display as '00:00' for midnight and '12:00' for noon, ensuring clarity and consistency across different regional settings.

SheikhSuhail19 replied:
FYI, while testing the suggested changes, I found that the time between 12:00:01 and 12:00:59 was not being handled correctly. I’ve addressed this by implementing truncation to the nearest minute, which ensures that any time between 12:00:00 and 12:00:59 (including 12:00:01) is correctly identified as Midnight or Noon.

This should now provide consistent and accurate labeling for times around midnight and noon.

SheikhSuhail19 replied:
Latest build:
![Uploading image.png…]()

Commit:
Update time formatting logic to only replace ambiguous 12:00 AM/PM labels (with "Midnight"/"Noon") when the user's locale uses a 12-hour clock.
For 24-hour formats, retain the standard "00:00"/"12:00" representation since these are unambiguous.

SheikhSuhail19 commented on core/src/main/resources/hudson/widgets/HistoryWidget/entry.jelly:
```diff
@@ -52,7 +52,7 @@ THE SOFTWARE.
           </j:if>
           <div tooltip="${linkTitleAttr}" data-tooltip-append-to-parent="true">
-            <i:formatDate value="${build.timestamp.time}" type="time" timeStyle="short" />
+            ${h.getDisambiguatedTime(build)}
```
Integrating localized terms directly within the FormatDateTag could indeed be complex.
I will explore this strategy further to assess its feasibility and effectiveness.

For now we can proceed with the current implementation of `getDisambiguatedTime`, which sufficiently addresses the time format disambiguation.

SheikhSuhail19 replied:
Latest build:
![image](https://github.com/user-attachments/assets/03cc0c76-9efd-4acc-9c96-8bac380a0c40)

A1exKH reviewed (approved):
LGTM.

mawinter69 commented on core/src/main/resources/hudson/Messages_de.properties:
```diff
@@ -75,3 +75,6 @@ AboutJenkins.DisplayName=Über Jenkins
 
 Functions.NoExceptionDetails=Keine Details zum Ausnahmefehler
+
+DisambiguatedTime.Midnight=0:00 Uhr
+DisambiguatedTime.Noon=12:00 Uhr
```
Here you're using the 24h format. Correct would be
```
12:00 Mitternacht
12:00 Mittag
```
But most likely this is unreachable anyway

mawinter69 reviewed (commented):
Does it make sense to provide the 12h strings for noon and midnight for locales that use 24h format?
Seems you used google to translate the times as I doubt you're a native speaker of so many languages.

jglick commented:
IMO the current behavior is not worth adjusting. If you are accustomed to 12h time formats, you should be accustomed to knowing what 12:00 AM and 12:00 PM mean in standard format strings. If you find these meanings confusing, switch to a 24h format. Anyway this would only apply to a build which started at that exact minute, which is not particularly likely; even if you are using a crontab, use the `H` syntax and most builds will not start exactly on the hour.

krisstern commented:
Agreeing with what @jglick has just stated

NotMyFault commented:
Given the former rationale given by Jesse and no further activity by the creator, I'll go ahead and close the PR.
